### ISSUE TITLE
pcfdev scs fails to deploy. 

### ISSUE THREAD

**Issue by PiNh00fD** [vmware-archive/pcfdev#393]
Hi, 
Installed pcfdev-v1.2.0-darwin.tgz on MacBook Pro with 16GB ram (same error occurs on Linux Bionic Beaver with 16G ram, so Tried multiple OS-es). works as expecte, but when I deploy service scs, the deployment fails. Added the last part of the deployment log, since I have the feeling that the phrase "Service broker error: Could not find an available leader" has something to do with it. I have the same error when I try to bind the "spring music test app" to mysql service.
(Service broker error: Could not find an available leader from [https://10.144.0.8:8443]). Below the deploy-service scs log.

```
Organization 'p-spring-cloud-services' not found.       
Space 'instances' not found.    
Space 'p-spring-cloud-services' not found.      
App spring-cloud-broker-venerable not found     
App spring-cloud-broker not found       
App spring-cloud-broker not found       
App spring-cloud-broker-worker not found        
Service instance spring-cloud-broker-db not found       
Unexpected Response     
Response code: 502      
CC code:       0        
CC error code:  
Request ID:    358f67d4-112b-499f-7b5b-d5b9a7631289::082b9a30-0afa-47f2-93ea-36d3da80ca33       
Description:   {        
  "description": "Service broker error: Could not find an available leader from [https://10.144.0.13:8443]",    
  "error_code": "CF-ServiceBrokerBadResponse",  
  "code": 10001,        
  "http": {     
    "uri": "http://10.144.0.7:8080/v2/service_instances/828aa933-a3f4-48bd-b254-367ca165374a/service_bindings/4ce8d30a-ef09-433a-b51a-3c120b7b7f11",    
    "method": "PUT",    
    "status": 500  
```

**Comment by jeksopra** [comment]
Hi together,
I get the same error when starting cf dev on a Windows 10 machine with the following start command:
```
cf dev start -f pcfdev-v1.2.0-windows.tgz -m 12288 -s scs
```

The issue also pops up if I try to deploy the SpringCloudServices manually after starting the cfdev without the '-s scs' option:
```
PS C:\Software\PCF_DEV> cf dev deploy-service scs
Deploying Spring Cloud Services...
  Progress: 1 of 4 (5m12s)FAILED
Failed to deploy services: Failed to deploy Spring Cloud Services: exit status 1
```

Then I tried to find out the cause by login to the AppManager and inspecting the particular application deployments of the org=system and space=p-spring-cloud-services. I determined that the Spring-Cloud-Broker app was down and not running. So I tried to restart this app, but it didn´t got started successfully. At reviewing the app specific logs in the AppManager I detected the following error signaling the ApplicationContext could not be instantiated sucessfully:

```
2020-06-15T10:31:37.156+02:00 [APP/PROC/WEB/0] [OUT] 2020-06-15 08:31:37.155 ERROR [spring-cloud-service-broker,,,] 23 --- [ost-startStop-1] o.s.b.web.embedded.tomcat.TomcatStarter : Error starting Tomcat context. Exception: org.springframework.beans.factory.BeanCreationException. 
Message: Error creating bean with name 'servletEndpointRegistrar' defined in class path resource 
[org/springframework/boot/actuate/autoconfigure/endpoint/web/
ServletEndpointManagementContextConfiguration$WebMvcServletEndpointManagementContextConfiguration.class]: 
Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: 
Failed to instantiate [org.springframework.boot.actuate.endpoint.web.ServletEndpointRegistrar]: 
Factory method 'servletEndpointRegistrar' threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: 
Error creating bean with name 'healthEndpoint' defined in class path resource [
org/springframework/boot/actuate/autoconfigure/health/HealthEndpointConfiguration.class]: 
Bean instantiation via factory method failed; nested exception is org.springframework.beans.BeanInstantiationException: 
Failed to instantiate [org.springframework.boot.actuate.health.HealthEndpoint]: 
Factory method 'healthEndpoint' threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: 
Error creating bean with name 'org.springframework.boot.actuate.autoconfigure.jdbc.DataSourceHealthIndicatorAutoConfiguration': 
Bean instantiation via constructor failed; nested exception is org.springframework.beans.BeanInstantiationException: 
Failed to instantiate [org.springframework.boot.actuate.autoconfigure.jdbc.DataSourceHealthIndicatorAutoConfiguration$$
EnhancerBySpringCGLIB$$b8b7b62d]: Constructor threw exception; nested exception is org.springframework.beans.factory.BeanCreationException: 
Error creating bean with name 'dataSource': Post-processing of FactoryBean's singleton object failed; 
nested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'scopedTarget.dataSource' defined 
in class path resource [org/springframework/boot/autoconfigure/jdbc/DataSourceConfiguration$Hikari.class]: Initialization of bean failed; n
ested exception is org.springframework.beans.factory.BeanCreationException: Error creating bean with name 
'org.springframework.boot.autoconfigure.jdbc.DataSourceInitializerInvoker': Invocation of init method failed; 
nested exception is org.springframework.jdbc.datasource.init.UncategorizedScriptException: Failed to execute database script; 
nested exception is org.springframework.beans.factory.BeanCurrentlyInCreationException: Error creating bean with name 'scopedTarget.dataSource': 
Requested bean is currently in creation: Is there an unresolvable circular reference?
2020-06-15T10:31:37.208+02:00 [APP/PROC/WEB/0] [OUT] 2020-06-15 08:31:37.207 INFO [spring-cloud-service-broker,,,] 23 --- [ main] o.apache.catalina.core.StandardService : Stopping service [Tomcat]
2020-06-15T10:31:37.217+02:00 [APP/PROC/WEB/0] [OUT] 2020-06-15 08:31:37.216 WARN [spring-cloud-service-broker,,,] 23 --- [ main] ConfigServletWebServerApplicationContext : Exception encountered during context initialization - cancelling refresh attempt: org.springframework.context.ApplicationContextException: Unable to start web server; nested exception is org.springframework.boot.web.server.WebServerException: Unable to start embedded Tomcat
2020-06-15T10:31:37.239+02:00 [APP/PROC/WEB/0] [OUT] Error starting ApplicationContext. To display the conditions report re-run your application with 'debug' enabled.
2020-06-15T10:31:37.239+02:00 [APP/PROC/WEB/0] [OUT] 2020-06-15 08:31:37.239 INFO [spring-cloud-service-broker,,,] 23 --- [ main] ConditionEvaluationReportLoggingListener :
2020-06-15T10:31:37.243+02:00 [APP/PROC/WEB/0] [OUT] └─────┘
2020-06-15T10:31:37.243+02:00 [APP/PROC/WEB/0] [OUT] | org.springframework.boot.autoconfigure.jdbc.DataSourceInitializerInvoker
2020-06-15T10:31:37.243+02:00 [APP/PROC/WEB/0] [OUT] ↑ ↓
2020-06-15T10:31:37.243+02:00 [APP/PROC/WEB/0] [OUT] | scopedTarget.dataSource defined in class path resource [org/springframework/boot/autoconfigure/jdbc/DataSourceConfiguration$Hikari.class]
2020-06-15T10:31:37.243+02:00 [APP/PROC/WEB/0] [OUT] ┌─────┐
2020-06-15T10:31:37.243+02:00 [APP/PROC/WEB/0] [OUT] dataSource
2020-06-15T10:31:37.243+02:00 [APP/PROC/WEB/0] [OUT] ↓
2020-06-15T10:31:37.243+02:00 [APP/PROC/WEB/0] [OUT] org.springframework.boot.actuate.autoconfigure.jdbc.DataSourceHealthIndicatorAutoConfiguration
2020-06-15T10:31:37.243+02:00 [APP/PROC/WEB/0] [OUT] healthEndpoint defined in class path resource [org/springframework/boot/actuate/autoconfigure/health/HealthEndpointConfiguration.class]
2020-06-15T10:31:37.243+02:00 [APP/PROC/WEB/0] [OUT] servletEndpointRegistrar defined in class path resource [org/springframework/boot/actuate/autoconfigure/endpoint/web/ServletEndpointManagementContextConfiguration$WebMvcServletEndpointManagementContextConfiguration.class]
2020-06-15T10:31:37.243+02:00 [APP/PROC/WEB/0] [OUT] **The dependencies of some of the beans in the application context form a cycle:**
2020-06-15T10:31:37.243+02:00 [APP/PROC/WEB/0] [OUT] Description:
2020-06-15T10:31:37.243+02:00 [APP/PROC/WEB/0] [OUT] ***************************
2020-06-15T10:31:37.243+02:00 [APP/PROC/WEB/0] [OUT] APPLICATION FAILED TO START
2020-06-15T10:31:37.243+02:00 [APP/PROC/WEB/0] [OUT] 2020-06-15 08:31:37.242 ERROR [spring-cloud-service-broker,,,] 23 --- [ main] o.s.b.d.LoggingFailureAnalysisReporter :
```
I hope this helps fixing the problem. 

**Since the CF DEV platform cannot currently be used in conjunction with the Spring Cloud Services at all (this also applies the former release pcfdev-v1.1.0-windows.tgz) I would very appreciate if you could take care about this issue.**

Thanks in advance and best regards!

**Comment by aemengo** [comment]
Hey @jeksopra Running Spring Cloud Services locally is a very resource intensive operation, requiring both the rabbitmq and mysql service to be deployed along with it. The usual problem that I've seen with this is that one's machine is limited is some way.

That said can you please post the output of `~/.cfdev/logs/deploy-spring-cloud-services.log`?

**Comment by jeksopra** [comment]
Hi @aemengo ,
thanks for your quick response. Hm - I startet the cf dev environment on an i7 Intel CPU with a dedicated memory amount of 12288m (cf dev start -m 12288m -s scs) and sometime even with 15000m (cf dev start -m 15000m -s scs). In the past I was successfully able to start the cf dev including the Spring Cloud Services in that way. The last time I experimented with it should be 5 month ago. There was no Issue starting and provisioning the SpringCloudServices. Please also have a look at my earlier post with the cycle failure at starting the Spring-Cloud-Broker application. 

By the way: Deploying the services mysql, redis and rabbitmq (with: cf dev deploy-service <service-name>) works very fine - also if you deploy all of them togehter. 

You can find the requested Logfile as a attachment to this comment.
Thanks a lot in advance!

[deploy-spring cloud services.log](https://github.com/pivotal-cf/pcfdev/files/4781141/deploy-spring.cloud.services.log)

**Comment by PiNh00fD** [comment]
I can confirm that the deploy-service works fine for rabbitmq and mysql.
@jeksopra, did you also try to create a mysql service and bind an app to it?

I deployed mysql and creating a service with cf create-service for the test app spring-music, but didn't bind it at first, still everything seems well:

Deployment 'pivotal-mysql-057f6d53046d9f108cb5'

Instance                                                     Process State  AZ    IPs         VM CID                                VM Type  Active  Stemcell  
dedicated-mysql-broker/1977b5c6-5b5e-464f-b38a-9faadbf82ea8  running        null  10.144.0.7  1923c187-a20e-4139-866b-7669c1d535cb  default  true    -  

1 vms

Deployment 'service-instance_bb4772c8-8d84-418e-ade5-853b36512a3f'

Instance                                    Process State  AZ    IPs         VM CID                                VM Type  Active  Stemcell  
mysql/4d45a001-76d8-43d5-93a8-4d574a731632  running        null  10.144.0.8  648031f4-5148-4b10-a9a2-07fe82e954d1  default  true    

Also when I do cf services, it looks ok. (or could it be the outdated api be an issue @aemengo)

ajh@ajh-OMEN-by-HP-Laptop:~/workspace/apps/spring-music$ cf services
Your CF API version (2.125.0) is no longer supported. Upgrade to a newer version of the API (minimum version 2.128.0). Please refer to https://github.com/cloudfoundry/cli/wiki/Versioning-Policy#cf-cli-minimum-supported-version

Getting services in org apps / space music as admin...

name    service   plan       bound apps   last operation     broker                   upgrade available
my-db   p.mysql   db-small                create succeeded   dedicated-mysql-broker   

TIP: Please upgrade to CC API v2.138.0 or higher for individual service upgrades

But.. When I want to bind the created service my-db to spring-music, it fails.

ajh@ajh-OMEN-by-HP-Laptop:~/workspace/apps/spring-music$ cf bind-service spring-music my-db
Your CF API version (2.125.0) is no longer supported. Upgrade to a newer version of the API (minimum version 2.128.0). Please refer to https://github.com/cloudfoundry/cli/wiki/Versioning-Policy#cf-cli-minimum-supported-version

Binding service my-db to app spring-music in org apps / space music as admin...
Service broker error: Could not find an available leader from [https://10.144.0.8:8443]

And also here, I see the same error as I see in the log of the deploy-service scs.

"Service broker error: Could not find an available leader from [https://10.144.0.13:8443]",    
  "error_code": "CF-ServiceBrokerBadResponse",  .

Could there be a relation, since both services are using the internal mysql DB,

Also, although I'm not an expert, leader and follower, assumes that there are 2 instances?
There is only one.

Hope this will help.

Perhaps some other usefull information:

ajh@ajh-OMEN-by-HP-Laptop:~/workspace/apps/spring-music$ cf -v
cf version 6.51.0+2acd15650.2020-04-07
ajh@ajh-OMEN-by-HP-Laptop:~/workspace/apps/spring-music$ cf api
api endpoint:   https://api.dev.cfdev.sh
api version:    3.60.0

**Comment by PiNh00fD** [comment]
I hope it reopens when I add a comment. Sorry!!

**Comment by jeksopra** [comment]
I can recreate what @PiNh00fD has mentioned about the SpringMusicApp when trying to bind an already successfully deployed mysql service to this app.
I got the follwing output:
PS> cf bind-service spring-music p-mysql-service
Binding service p-mysql-service to app spring-music in org system / space system as admin...
Service broker error: Could not find an available leader from [https://10.144.0.8:8443]  

After that I tried the same with the latest **CF DEV** edition (**not** PCF DEV it is important to differenciate between both). There deploying the mysql-service with ' cf create-service p-mysql 20mb p-mysql-service' and binding the Spring Music App to the aforementioned mysql-service with 'cf bind-service spring-music p-mysql-service' worked fine. 
After restaging the Spring-Music-App it was possible to store new data records permanently in the service-backed MySQL database instance.

**Comment by PiNh00fD** [comment]
Any updates on this issue?

**Comment by slindenberg** [comment]
Same here. MySQL and RabbitMQ can be installed, SCS fails ("progrss 1 of 4"). The start parameters where: `cf7 dev start -c 2 -m 16384 -f /pcfdev-v1.2.0-linux.tgz`, starts without errors. Then `cf7 dev deploy-service scs`, which fails. Thanks for any support!

**deploy-springcloud-services.log:**
[deploy-spring cloud services.log](https://github.com/pivotal-cf/pcfdev/files/4831344/deploy-spring.cloud.services.log)

**Comment by jdegaul803** [comment]
Same here mysql, rabbitmq , redis deploy fine, scs fails both using the -s scs and deploy-service. 
Machine : MacOS catalina
Cf version: 6.51.0+2acd15650.2020-04-07. 
log: 
[deploy-spring cloud services.log](https://github.com/pivotal-cf/pcfdev/files/4841907/deploy-spring.cloud.services.log)
thanks in advance

**Comment by PiNh00fD** [comment]
I think, there is a new issue opened for this topic.

https://github.com/cloudfoundry-incubator/cfdev/issues/132

**Comment by aemengo** [comment]
@PiNh00fD @jeksopra @slindenberg @jdegaul803  

I'd like to sincerely apologize to everyone on the thread. After continued investigation, I able to see that there was an internal error with CF Dev. At the point of its release, there was not a concern but the internal certificates used in the Spring Cloud Services execution path have expired.

```sh
# /var/vcap/sys/log/broker/broker.log
[on-demand-service-broker] [f347182b-3834-4f1d-9648-d92ebbe8fb51] 2020/06/30 22:47:58.061686 service adapter will delete binding with ID b3756eda-c00f-42c2-b9fa-dbb78fd06ddd for instance 0b252b61-0831-4cf3-a920-bd68dc012a20
[on-demand-service-broker] [f347182b-3834-4f1d-9648-d92ebbe8fb51] 2020/06/30 22:47:58.088705 external service adapter exited with 1 at /var/vcap/packages/odb-service-adapter/bin/service-adapter:
stdout:
    'Could not find an available leader from [https://10.144.0.13:8443]',
stderr: '
    [odb-sdk] handling delete-binding
    [dedicated-mysql-adapter] 2020/06/30 22:47:58 Using agent deleter
    [dedicated-mysql-adapter] 2020/06/30 22:47:58 Couldn't get read only information about URL https://10.144.0.13:8443, error: &errors.errorString{s:"GET https://10.144.0.13:8443/status failed: Get https://10.144.0.13:8443/status: x509: certificate has expired or is not yet valid"}
    [odb-sdk] Could not find an available leader from [https://10.144.0.13:8443]
    '
```
To rectify, I just released updates for the CF Dev product that will have newer certificates. This means that the problem is abated but, of course, not resolved forever.

**New Release**
- **v0.0.18** (plugin) - https://plugins.cloudfoundry.org/#cfdev
- ~~**v1.3.0** (Service Assets) - https://network.pivotal.io/products/pcfdev/#/releases/685614~~
- **v1.3.1** (Service Assets) - https://network.pivotal.io/products/pcfdev#/releases/690049

-----

I'd also like to remind everyone that this is a repository for a _dead product_ and has even less eyes on it than its replacement: https://github.com/cloudfoundry-incubator/cfdev. For the sake of maintenance, please refrain from opening issues here.
